### ISSUE TITLE
Add Native Payments

### ISSUE THREAD

**Issue by dengjeffrey** [Shopify/unity-buy-sdk#65]
## What
Allow the SDK to support Apple Pay and Android Pay as part of the checkout flow.

## How
Integrating native in app payments require the use of [Native Plugins](https://docs.unity3d.com/Manual/Plugins.html)

+ **TLDR:** 
	+ Accessing Plugin functions from Unity: 
		+ On iOS we expose functions to Unity by wrapping methods in C++, on Android we will access the JVM and query for class and method definitions.
	+ Accessing Unity functions from Plugins:
		+ `UnitySendMessage("GameObjectName1", "MethodName1", "Message to send")`
		+ The message must be a UTF8 String 

## Checklist

- [X] Serialize `PassKit` Objects going to Unity
- [x] Deserialize `PassKit` Objects coming from Unity
- [x] `PaymentSession` that manages the Authorization Controller
- [x] `ApplePayEventDispatcher` that forwards events from Authorization Controller to Unity
- [x] `MessageCenter` that handles responses from Unity
- [x]   `Cart+ApplePay` a native C++ wrapper that exposes methods to create a `PaymentSession`
- [x] Shipping mutations for checkout query on Unity
- [x] Hooks on `Cart` that allow the user to create a `PaymentSession`
- [x] Handle messages from the `ApplePayEventDispatcher` in Unity
- [x] Apple Pay
- [ ] Android Pay

**Comment by dengjeffrey** [comment]
Below is my thought process and possible solution to implementing this. I have not experimented on Android thoroughly so any input will be helpful

## How
### iOS
+ On iOS I have considered implementing Apple Pay in Objective C, but since Objective C is not likely to be maintained in the future and we do not lose any functionality by implementing it in Swift, I went with the latter.
+ By using a bridging header we will expose Swift functions to an ObjC++ wrapper.
+ Using a ObjC++ wrapper we expose the functions to Unity scripts.

### Android
+ There seems to be few number of ways to implement Android plugins each with there own caveats. TBD 

## Proposed Native Payment Flow

1. On a checkout screen we will make a call `Cart.canCheckoutWithNativePay() -> Bool` which determines if Apple Pay or Android Pay is available on the device. Based on the result the we will show a Native Pay button.
	+ If `false` is returned on an iOS device, we can decide if we would like to support an integrated Apple Pay Setup using `Cart.canShowApplePaySetup() -> Bool` to determine if the device supports setting up a new card. If so we can display a setup Apple Pay Button and notify the Swift plugin when it is pressed.

2. Once a Native Pay button is pressed, we will call either `Cart.checkoutWithApplePay(string merchantID)` or `Cart.prepareCheckoutWithAndroidPay(string publicKey)` to continue the checkout process.
	+ The flow must be split into separate device specific functions because paying with Apple Pay is a one step process, where as Android Pay requires two steps as described below.

	#####Android
	1.  `prepareCheckoutWithAndroidPay()` tells the Android Plugin to create a [`MaskedWalletRequest`](https://developers.google.com/android/reference/com/google/android/gms/wallet/MaskedWalletRequest) and fetch a [`MaskedWallet`](https://developers.google.com/android/reference/com/google/android/gms/wallet/MaskedWallet). This is where the user chooses a payment option and fills out their shipping address. When a `MaskedWallet` is returned, the Android Plugin will pass a serialized `MaskedWallet` to Unity. We will display the `MaskedWallet` info to the user so that they can confirm that the shipping address and payment card is correct.
		+ We do not use a [`WalletFragment`](https://developers.google.com/android/reference/com/google/android/gms/wallet/fragment/WalletFragment) to facilitate modifications to the `MaskedWallet` because we can not specify where to place a `WalletFragment` without having to subclass [`UnityPlayerActivity`](https://github.com/eppz/Unity.Labs.Unity_Player_Android/blob/master/com/unity3d/player/UnityPlayerActivity.java) which is adding more restrictions for the implementor. Instead we will provide a hook in the Android Plugin to access [`changeMaskedWallet()`](https://developers.google.com/android/reference/com/google/android/gms/wallet/Payments.html#changeMaskedWallet(com.google.android.gms.common.api.GoogleApiClient, java.lang.String, java.lang.String, int))
		+ The `MaskedWallet` is stored in the Android Plugin for future requests when we create a [`FullWallet`](https://developers.google.com/android/reference/com/google/android/gms/wallet/FullWallet)

		```c#
		// Unity
		void prepareCheckoutWithAndroidPay(string publicKey) {
			// _Function denotes a function exposed from our Android Plugin
		
			// Android Plugin will create a MaskedWalletRequest and fetch a MaskedWallet
			_FetchMaskedWallet(publicKey, Store.name, Store.currencyCode);
			
			// When the Android Plugin recieves the MaskedWallet
			// it will call DidFetchMaskedWallet
		} 
		
		void DidFetchMaskedWallet(string message) {
			// Parse message, and display info from MaskedWallet
			... 
		}
		```
		
	2. Once the user has pressed the confirm button we will call `checkoutWithAndroidPay()` which tells the Android Plugin to create a  `FullWalletRequest` and fetch a `FullWallet`
		+ The process for this is similar to fetching a `MaskedWallet`
	3. The payment token from `FullWallet` is passed to Unity where we will complete the token checkout

		```c#
		// Unity
		void CompleteCheckoutWithToken(string message) {
			... 
		}
		```

	#####iOS
	1. `checkoutWithApplePay(string merchantID)` consists of 3 sub steps handled by the Swift Plugin.
		1. Create a [`PKPaymentRequest`](https://developer.apple.com/reference/passkit/pkpaymentrequest?language=objc)
		2. Create [`PKSummaryItem`](https://developer.apple.com/reference/passkit/pkpaymentsummaryitem?language=objc) from [`CartLineItem`](https://github.com/Shopify/unity-buy-sdk/pull/64) and other Cart details
		3. Present [`PKPaymentAuthorizationViewController`](https://developer.apple.com/reference/passkit/pkpaymentauthorizationviewcontroller?language=objc) on top of the current `ViewController`

		```c#
		// Unity
		void checkoutWithApplePay(string merchantID) {
			// _Function denotes a function exposed from a C# wrapper of ObjC/Swift methods
		
			// Create a PKPaymentRequest
			_StartPaymentRequest(merchantID, Store.countryCode, Store.currencyCode);
		
			// Create PKSummaryItems
			_AddSummaryItem("Cart Total", Cart.total());
			_AddSummaryItem("Shipping Rate", Cart.shippingRate());
			_AddSummaryItem("Taxes", Cart.taxes());
			...
		
			// Present PKPaymentAuthorizationViewController
			_ProcessPaymentRequest();
		} 
		```
		+ Changes to the shipping address and payment will be handled by the`PKPaymentAuthorizationViewControllerDelegate` in the Swift Plugin. 
			+ When the delegate is called to update the summary items. We will send the updated information to Unity and block till summary items are updated. We need to block because calls to Unity are asynchronous, so we must wait till the we have recalculated the cart.
		
		```swift
		// Swift Plugin
		func paymentAuthorizationViewController(PKPaymentAuthorizationViewController, 
				       didSelect shippingMethod: PKShippingMethod,
							     	 completion: (PKPaymentAuthorizationStatus, [PKPaymentSummaryItem]) -> Void)
		{
			// Notify Unity of a change, this call is asynchronous
			UnityWrapper.sendMessage(shippingMethod.identifier, toObject:"CartObject", withMethodName:"UpdateShippingForIdentifier")
			
		    updateSemaphore.wait()
			completion(summaryItems)
		}
	
		func didFinishUpdatingSummaryItems() {
			updateSemaphore.signal()
		}		
		```
		
		```c#
		// Unity
		void UpdateShippingForIdentifier(string message) {
		
			// Calculate new shipping rate here
			... 
			
			// Update the summary items and notify Swift Plugin
			_AddSummaryItem("Shipping Rate", Cart.shippingRate)	
			_DidFinishUpdatingSummaryItems(); 
		}
		```
			
	2. The token returned from `PKPaymentAuthorizationViewControllerDelegate` is passed to Unity where we will complete the token checkout


Cancellations of the payment during the process should be relatively trivial to add



##Issues Encountered
+ Calls to `UnitySendMessage` must be on the heap in Swift, I am currently exploring ways to make call to `UnitySendMessage` in `@escaping` closures.

**Comment by dbart01** [comment]
I'm not sure there's any benefit (aside from syntax) in using Swift. Since all classes defined in Swift have to stay compatible with ObjC (can't use `rawValue` backed `enum`, can't use generics, etc), there's really not much benefit in using Swift compared to just Objective-C. What you do gain with ObjC though is direct compatibility with Objective-C++, just rename to `.mm` and you're set. Whether the nice Swift syntax is worth the trade-off with C++ compatibility if your call but just something to keep in mind. It might eliminate some boiler-plate wrappers.

**Comment by mikkoh** [comment]
This looks pretty good! I think we still need to think about this. After reading this it’s still unclear to me how this would fully work. Here are some questions and comments I had when reading through this.


## Questions
* I’m unsure of the decision to use Swift at this point. Mainly because I don’t know the mobile ecosystem well. It would be a huge bummer if a developer was attempting to incorporate the Unity SDK into a pre-existing published project and was not able to use Swift for some reason or another.  @davidmuzi @dbart01 do you see this as an issue? Knowing that Unity generates Xcode projects fully which can compile as a native app is this still an issue?
* `Cart.CanShowApplePaySetup()` didn’t even think of this. This is great. Can we do this for Android also. If we can we could do `Cart.CanShowNativePaySetup()`
* Where does `publicKey` come from for `void PrepareCheckoutWithAndroidPay(string publicKey) {`?
* For `void CompleteCheckoutWithToken(string message)` should it be `void CompleteCheckoutWithToken(string token)`?
* Where does `merchantId` come from for `void CheckoutWithApplePay(string merchantID) {`?
* When you state _*We will display the MaskedWallet info to the user so that they can confirm that the shipping address and payment card is correct.*_ does this mean that we need to build UI that shows this info?
* Does `_FetchMaskedWallet` bring up native ui? Is the user able to cancel this process? `CompleteCheckoutWithToken` is called on success but is there a failure case?
* What are the failure cases for iOS? Can the user cancel the payment? Can it fail in another way?
* What data fields/data do we need to query before we can attempt to checkout on different platforms?
	* Android: Store Name, Store Currency? Anything else?
	* iOS: Shipping Methods, Tax Info, Cart Total (before taxes I assume)? Anything else?

## Comments
I still would like to explore if it’s possible to consolidate the steps so that the process of native checkout is similar. Could we do something like this?

* `cart.CanCheckoutWithNativePay()`
* `cart.CanShowNativePaySetup()`
* `cart.PrepareNativeCheckout(OnPrepareSucess, OnPrepareFailure)` (I know on iOS this would not do much but I'd rather keep the process consistent than efficient)
* `cart.CompleteCheckoutWithToken(token, OnPrepareSucess, OnPrepareFailure)`

**Comment by dengjeffrey** [comment]
### Questions Answered
+ I’m unsure of the decision to use Swift at this point. Mainly because I don’t know the mobile ecosystem well. It would be a huge bummer if a developer was attempting to incorporate the Unity SDK into a pre-existing published project and was not able to use Swift for some reason or another. 
    + Since Apple Pay requires iOS 8+, Swift will be available also. @dbart01  raises a good point, but we would still have to wrap ObjC in C++ methods for Unity and so far it seems like we only need to wrap `UnitySendMessage` from C++ to Swift. The positive trade off is we would be building for the long term, for if/when Swift has a higher adoption rate.

+ Cart.CanShowApplePaySetup() didn’t even think of this. This is great. Can we do this for Android also. If we can we could do Cart.CanShowNativePaySetup()
    + As far as I can tell from the docs Android does not have functionality for setting up Android Pay whilst in the app. They suggest showing the user a message to set up Android Pay.

+ Where does publicKey come from for void PrepareCheckoutWithAndroidPay(string publicKey) {?
    + The [`publicKey`](https://developers.google.com/android-pay/integration/payment-token-cryptography) is provided by the implementor of the SDK. The key is used by Android to encrypt wallet information, and later on we use it to decrypt the `FullWallet` in order to access the 
Payment Token.

+ For void CompleteCheckoutWithToken(string message) should it be void CompleteCheckoutWithToken(string token)?
    + Yes should be `void CompleteCheckoutWithToken(string token)` good catch

+ Where does merchantId come from for void CheckoutWithApplePay(string merchantID) {?
    + `merchantId` is provided by the implementor of the SDK. When creating an app with Apple Pay capabilities, they must create a `merchantId` from the Apple Developer Portal.

+ When you state We will display the MaskedWallet info to the user so that they can confirm that the shipping address and payment card is correct. does this mean that we need to build UI that shows this info?
    + We should not need to build this UI. But the implementor of this SDK should build a UI in Unity that does display the shipping address and payment card, or they can use the `WalletFragment` which is a fragment provided by the Android API. If they choose to use a `WalletFragment` then they would have to subclass or modify the behaviour of the `UnityPlayerActivity` to display the `WalletFragment`. I think we should leave this up to the implementor because they may already have subclassed the UnityPlayerActivity which may cause conflicts. This part of the flow is shown in step 3 of the [Android Payment Flow](https://developers.google.com/android-pay/payment-flows)

+ Does _FetchMaskedWallet bring up native ui? Is the user able to cancel this process? CompleteCheckoutWithToken is called on success but is there a failure case?
    + ` _FetchMaskedWallet` and `_FetchFullWallet` will bring up native UI, and they can cancel it. If `_FetchMaskedWallet` results in an error or if the user cancels, then this error message will be propagated to Unity via `UnitySendMessage` where we can notify the implementor via a delegate message or invoke a Unity`Action` closure to have the Implementor show a customized alert. We can also provide a default option and present a native toast similar to how Mobile Buy SDK - Android  handles it:

```java
    private void handleGoogleWalletError(int errorCode) {
        switch (errorCode) {
            // Recoverable Wallet errors
            case WalletConstants.ERROR_CODE_SPENDING_LIMIT_EXCEEDED:
                Log.e(LOG_TAG, "Spending limit exceeded");
                Toast.makeText(this, "Spending Limit exceeded. Please adjust your cart and try again", Toast.LENGTH_SHORT).show();
                launchProductListActivity();
                break;

            // Unrecoverable Wallet errors
            case WalletConstants.ERROR_CODE_INVALID_PARAMETERS:
            case WalletConstants.ERROR_CODE_AUTHENTICATION_FAILURE:
            case WalletConstants.ERROR_CODE_BUYER_ACCOUNT_ERROR:
            case WalletConstants.ERROR_CODE_MERCHANT_ACCOUNT_ERROR:
            case WalletConstants.ERROR_CODE_SERVICE_UNAVAILABLE:
            case WalletConstants.ERROR_CODE_UNSUPPORTED_API_VERSION:
            case WalletConstants.ERROR_CODE_UNKNOWN:
            default:
                Log.e(LOG_TAG, "Unrecoverable wallet error:" + errorCode);
                Toast.makeText(this, "Could not complete the checkout, please try again later", Toast.LENGTH_SHORT).show();
                break;
        }
    }
```

+ What are the failure cases for iOS? Can the user cancel the payment? Can it fail in another way?
    + For iOS the user can cancel the payment in which case we wouldn't need to do anything, but to dismiss `PKPaymentAuthorizationViewController`. Since the Apple Pay flow requires that we complete the checkout by sending necessary payment data to the server before dismissing the `PKPaymentAuthorizationViewController` we can fail on the call to checkout, which `PKPaymentAuthorizationViewController` will handle by displaying the status message to the user.

```swift
func paymentAuthorizationViewController(_ controller: PKPaymentAuthorizationViewController, didAuthorizePayment payment: PKPayment, completion: @escaping (PKPaymentAuthorizationStatus) -> Void) {
  let authStatus = ... status from performing checkout 
   completion(authStatus)
}
``` 

+ What data fields/data do we need to query before we can attempt to checkout on different platforms?
    + Android Specific
        + Public Key (Provided by the implementor)
        + Shop name
        + Countries that Merchant Ships to
    + iOS Specific
       + Merchant ID (Provided by the implementor)
       + Shop country
    + Both platforms
      + Does checkout require shipping
      + Currency Code
      + Shipping rates
      + Discounts 

I'll also add some of these comments above. I think you're right making it consistent is probably less confusing for the implementor. @mikkoh

**Comment by mikkoh** [comment]
I'll try to read over this comment again and give detailed feedback. But I just want to bring up one thing.

```
we can fail on the call to checkout, which PKPaymentAuthorizationViewController will handle by displaying the status message to the user
```

I think it would be good if we still gave the Unity developer the ability to be reactive to a failure. For instance lets say in their game the main character reacts to users decision to not buy product. For instance lets say they shake their fist towards the screen when the user doesnt buy the stuff their selling. Or gives a thumbs up after the purchase experience if they did buy something.

**Comment by mikkoh** [comment]
@dengjeffrey any updates on this? You mentioned today that @dbart01 had given you some new ideas on implementation.

**Comment by dengjeffrey** [comment]
@mikkoh I will have a more detailed write up before noon tomorrow

**Comment by mikkoh** [comment]
@dengjeffrey here are some updated questions and me just trying to think all of this through.

- Idea! Can we do `CanShowNativePaySetup` instead of `CanShowApplePaySetup`. `Cart.CanShowNativePaySetup()` always returns `false` for Android this means that the developer can write code once for all platforms.
- For Android is `key` just generated? If so is there any value allowing the user to pass in their own key or do we just generate once. If we make this optional how would that effect iOS's method signature.
- On Success for `PrepareNativeCheckout` what would the user use `WalletInfo` for? What will be included in `WalletInfo`? 

Let me know if this example usage makes sense:
```cs
if (Cart.CanCheckoutWithNativePay()) {
    Cart.PrepareNativeCheckout(key, (walletInfo) => {
        // what can I do with walletInfo ?!?

        Cart.CompleteNativeCheckout(onCompleteSucess, onCompleteFail);
    }, onPrepareFail);
} else if (Cart.CanShowNativePaySetup()) {
    Cart.ShowNativePaySetup(onNativePaySetupSuccess, onNativePaySetupFail);
}
```

All methods on Cart for native checkout:
```cs
Cart.CanCheckoutWithNativePay();
Cart.CanShowNativePaySetup();
Cart.ShowNativePaySetup();
Cart.PrepareNativeCheckout(key, onPrepareSuccess, onPrepareFail);
Cart.CompleteNativeCheckout(onCompleteSucess, onCompleteFail);
```

**Comment by dengjeffrey** [comment]
+ Yes we can have `CanShowNativePaySetup()` return false for Android, it makes the flow more simple but I hope it won't cause confusion. 

+ `key` is a public key from the implementor, this public key is generated on Shopify's server for the store. They get this key from the Mobile App sales channel on their store dashboard. 

+ walletInfo contains the following: 
    + `String[] paymentDescriptions `
        + [public String[] getPaymentDescriptions ()](https://developers.google.com/android/reference/com/google/android/gms/wallet/MaskedWallet) of MaskedWallet
        + An array of strings describing the payment method that the user chose, not exactly sure what this array of strings will contain, but it they should be shown to the user according to the Google Wallet UI guidelines
    + `MailingAddress shippingAddress` 

Below is what an example usage would look like. 

```c#
if (Cart.CanCheckoutWithNativePay()) {
    // Display Pay with Android or Apple Pay button, alongside web checkout button
    #if UNITY_IOS
    ApplePayButton.Instantiate(onClick: () => {
        Cart.PrepareNativeCheckout(key, (walletInfo) => {
            // Do nothing with walletInfo, since it's null on Apple Pay
            Cart.CompleteNativeCheckout(onCompleteSucess, onCompleteFail);   
        }
    })
    #elif UNITY_ANDROID
    AndroidPayButton.Instantiate(onClick: () => {
        Cart.PrepareNativeCheckout(key, (walletInfo) => {
            // Display a confirmation UI to the user based on the info WalletInfo, the implementor has to make this UI
            // walletInfo contains the following: 
            // String[] paymentDescriptions (https://developers.google.com/android/reference/com/google/android/gms/wallet/MaskedWallet)
            // MailingAddress shippingAddress 
            WalletConfirmationUI.Instantiate(onConfirmClick: () => {
                Cart.CompleteNativeCheckout(onCompleteSucess, onCompleteFail);   
            })
        }
    })
    #else 
    // Show some Unity editor button for testing
    #endif
} else if (Cart.CanShowNativePaySetup()) {
   // Display a Setup Apple Pay button. 
   // This part might cause confusion for the implementor
   // whether or not the implementor will need a Button for Android, 
   // they will have to pay close attention to our Docs.
    Cart.ShowNativePaySetup(onNativePaySetupSuccess, onNativePaySetupFail);
}
```
In addition to the methods you have mentioned in Cart for native checkout, we need to include Update methods. So far for iOS:

```c#
void UpdateSummaryItemsForShippingIdentifier(string identifier)
void UpdateSummaryItemsForShippingAddress(string addressJSON)
void UpdateSummaryItemsForShippingContact(string contactJSON)
void UpdateSummaryItemsForPaymentMethod(string paymentMethodJSON)
// This one is called when a payment token is generated via Apple Pay and 
// Apple pay needs a server status whether the payment went through or not before dismissing
void FetchCheckoutStatusForToken(string token)

// Calls to Plugin
bool CanCheckoutWithNativePay() 
bool CanShowNativePaySetup()
void ShowNativePaySetup()
void PrepareNativeCheckoutWithKey(string key, Action<WalletInfo> success, Action<CheckoutFailureType, string> failure)
void CompleteNativeCheckout(Action success, Action<CheckoutFailureType, string> failure) 
```
